### ISSUE TITLE
[Feature Request] Jagged batches support (NJT) for Transformer / SDPA inference

### ISSUE THREAD

**Issue by vadimkantorov** [microsoft/onnxruntime#22764]
### Describe the feature request

PyTorch / HF (previously branded as BetterTransformer) now have some support for NJT representation:
- https://github.com/onnx/onnx/issues/6525

This allows to have efficient inference in context of continuous batching.

Does ORT have such NJT-enabled SDPA kernels built-in?

Thanks!

### Describe scenario use case

N/A

**Comment by tianleiwu** [comment]
Please take a look at https://github.com/microsoft/onnxruntime/blob/e7987a6b0ba429c0bec248c4a471e1782da4be6c/onnxruntime/python/tools/transformers/notebooks/PyTorch_Bert-Squad_OnnxRuntime_GPU.ipynb

There is a section of "Packing Mode (Effective Transformer)".

**Comment by vadimkantorov** [comment]
Is there anywhere more documentation of this packing mode and the transformation into the packing mode? Is there any mapping onto NJT from PyTorch? (or examples of export to ONNX of NJT-supporting ops)

Does it ensure that positional embeddings restart from zero?

There is "ragged batching" support in Triton Inference Server https://github.com/triton-inference-server/server/blob/main/docs/user_guide/ragged_batching.md which lets the server itself concat inputs and add input of offsets/lengths. In what format does this BERT accept these extra inputs? Does it rely on BOS/EOS to separate batch elements?

**Comment by tianleiwu** [comment]
There is no detail document about that, and you can find the conversion script here: https://github.com/microsoft/onnxruntime/blob/4d614e15bd9e6949bc3066754791da403e00d66c/onnxruntime/python/tools/transformers/convert_to_packing_mode.py.

The packing mode follows the idea of [effective transformer](https://github.com/bytedance/effective_transformer). It assumes that your bert model has inputs with padding and use attention mask to indicates the paddings. Then the conversion script will add a node to remove paddings before the first Attention or MultiHeadAttention node, and insert another node to restore padding after the last LayerNormalization node. It does not have assumption that padding on the right side, and it can handle paddings at any position.

It is slightly difference from Triton ragged batching since Triton requires the inputs with right-side padding removed.

ORT is capable of support ragged batching since the Attention or MultiHeadAttention operator supports passing sequence lengths as mask index. I could add a ragged batch example later.

**Comment by vadimkantorov** [comment]
> I could add a ragged batch example later.

I think it would be very valuable!

Especially if it could also be a complete example with ONNX export of BERT from PyTorch and post-processing the ONNX file
